## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :x: [more...](#afda8c93-51c4-4257-8383-874141e67229) |
| mint-fs.sh | :x: [more...](#9baef0c5-3906-4508-8fc9-d2298b5eaeed) |
| mint-gateway-s3.sh | :x: [more...](#e29509aa-bb45-48d8-ba17-de2e26a320e6) |
| mint-erasure.sh | :x: [more...](#db1ed7e8-216b-4142-aeea-9cd34e9a338f) |
| mint-dist-erasure.sh | :x: [more...](#a1b37cd3-9e60-4d14-afa8-a9dfb89a83ba) |
| mint-zoned.sh | :x: [more...](#6ed43c93-53aa-486d-b181-705a0c554f74) |
| mint-gateway-nas.sh | :x: [more...](#32b0f55e-ccbf-4e3c-827b-498b68851ff0) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#9b5d22c9-ced1-4479-bce0-1c36d1ee7808) |
<a name="db1ed7e8-216b-4142-aeea-9cd34e9a338f"></a>
### 11801-394cc2c/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30319
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d48be5d88953:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 2 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... FAILED in 5 seconds
{
  "name": "mc",
  "duration": "4528",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-30849 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-17172/mc-test-object-11798  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-17172/mc-test-object-11798`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 3.50 MiB, Speed: 823.95 KiB/s"
}
(7/15) Running minio-dotnet tests ... done in 33 seconds
(8/15) Running minio-go tests ... FAILED in 22 seconds
{
  "args": {
    "bucketName": "minio-go-test-3uuoc5xmxvyvmi2i",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 4658,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 44 seconds
(9/15) Running minio-js tests ... FAILED in 3 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 4,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(9/15) Running minio-py tests ... done in 2 minutes and 7 seconds
(10/15) Running s3cmd tests ... done in 24 seconds
(11/15) Running s3select tests ... done in 4 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="afda8c93-51c4-4257-8383-874141e67229"></a>
### 11801-394cc2c/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30151
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3429db4d57f0:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 3 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 15 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... FAILED in 7 seconds
{
  "name": "mc",
  "duration": "5421",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-29338 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-12906/mc-test-object-22216  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-12906/mc-test-object-22216`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 9.75 MiB, Speed: 1.87 MiB/s"
}
(7/15) Running minio-dotnet tests ... done in 43 seconds
(8/15) Running minio-go tests ... FAILED in 25 seconds
{
  "args": {
    "bucketName": "minio-go-test-yhw2wr4el5ut54x9",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 4836,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 1 minutes and 28 seconds
(9/15) Running minio-js tests ... done in 46 seconds
(10/15) Running minio-py tests ... done in 3 minutes and 29 seconds
(11/15) Running s3cmd tests ... done in 18 seconds
(12/15) Running s3select tests ... done in 7 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="9baef0c5-3906-4508-8fc9-d2298b5eaeed"></a>
### 11801-394cc2c/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31330
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1b23ef2d1576:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 5 seconds
{
  "name": "mc",
  "duration": "4080",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-10 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-20666/mc-test-object-7308  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-20666/mc-test-object-7308`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 3.06 MiB, Speed: 809.27 KiB/s"
}
(7/15) Running minio-dotnet tests ... done in 31 seconds
(8/15) Running minio-go tests ... FAILED in 14 seconds
{
  "args": {
    "bucketName": "minio-go-test-q06ex94tl9wgb9xl",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 3166,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 21 seconds
(9/15) Running minio-js tests ... done in 42 seconds
(10/15) Running minio-py tests ... done in 1 minutes and 6 seconds
(11/15) Running s3cmd tests ... done in 14 seconds
(12/15) Running s3select tests ... done in 4 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="32b0f55e-ccbf-4e3c-827b-498b68851ff0"></a>
### 11801-394cc2c/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30475
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d112a587fbe7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 6 seconds
{
  "name": "mc",
  "duration": "4343",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-29847 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-5911/mc-test-object-26303  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-5911/mc-test-object-26303`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 3.38 MiB, Speed: 830.16 KiB/s"
}
(7/15) Running minio-dotnet tests ... done in 31 seconds
(8/15) Running minio-go tests ... FAILED in 17 seconds
{
  "args": {
    "bucketName": "minio-go-test-p0a2963vt6o32rrt",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 4592,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 10 seconds
(9/15) Running minio-js tests ... done in 31 seconds
(10/15) Running minio-py tests ... done in 59 seconds
(11/15) Running s3cmd tests ... done in 13 seconds
(12/15) Running s3select tests ... done in 3 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="9b5d22c9-ced1-4479-bce0-1c36d1ee7808"></a>
### 11801-394cc2c/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32240
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3628787228ad:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2924,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:32240 s3api copy-object --bucket awscli-mint-test-bucket-2443 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-2443/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... FAILED in 6 seconds
{
  "name": "mc",
  "duration": "4579",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-19518 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-12936/mc-test-object-12778  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-12936/mc-test-object-12778`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 9.81 MiB, Speed: 2.23 MiB/s"
}
(6/15) Running minio-dotnet tests ... done in 40 seconds
(7/15) Running minio-go tests ... FAILED in 18 seconds
{
  "args": {
    "bucketName": "minio-go-test-rbl5legfjn1qnko6",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 4749,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 58 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 182,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-3up4rhm, objectName = minio-java-test-1niltlk-copy, resource = /minio-java-test-3up4rhm/minio-java-test-1niltlk-copy, requestId = 166CE56C73239170, hostId = 555f726a-a92a-4f3a-b726-b2b0efb8979b)\nrequest={method=PUT, url=http://minio-dev4.minio.io:32240/minio-java-test-3up4rhm/minio-java-test-1niltlk-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-3vf0hub/minio-java-test-1niltlk\nHost: minio-dev4.minio.io:32240\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210316T181700Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210316/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 16 Mar 2021 18:17:00 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166CE56C73239170\nX-Xss-Protection: 1; mode=block\nDate: Tue, 16 Mar 2021 18:17:00 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 51 seconds
(8/15) Running minio-py tests ... done in 3 minutes and 18 seconds
(9/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2740",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-18179/s3cmd-test-object-5297 01b2e76b-8706-42af-b15a-c89b25963e7d\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 01b2e76b-8706-42af-b15a-c89b25963e7d put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(9/15) Running s3select tests ... done in 10 seconds
(10/15) Running security tests ... done in 0 seconds

Executed 10 out of 15 tests successfully.
```
<a name="6ed43c93-53aa-486d-b181-705a0c554f74"></a>
### 11801-394cc2c/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30816
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 25977b2527d3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 2 minutes and 9 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 7 seconds
{
  "name": "mc",
  "duration": "4929",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-5562 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-1704/mc-test-object-15184  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-1704/mc-test-object-15184`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 10.00 MiB, Speed: 2.11 MiB/s"
}
(7/15) Running minio-dotnet tests ... done in 46 seconds
(8/15) Running minio-go tests ... FAILED in 24 seconds
{
  "args": {
    "bucketName": "minio-go-test-kkr09vtx4hv9son3",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 3979,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 1 minutes and 35 seconds
(9/15) Running minio-js tests ... done in 47 seconds
(10/15) Running minio-py tests ... done in 3 minutes and 7 seconds
(11/15) Running s3cmd tests ... done in 18 seconds
(12/15) Running s3select tests ... done in 6 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="e29509aa-bb45-48d8-ba17-de2e26a320e6"></a>
### 11801-394cc2c/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30131
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b681cdcc4762:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-5041h169y5edpn",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 781,
  "error": "InternalError: We encountered an internal error, please try again.: cause(malformed chunked encoding)\n\tstatus code: 500, request id: 166CE53CC6F77322, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go PUT expected to success but got InternalError: We encountered an internal error, please try again.: cause(malformed chunked encoding)\n\tstatus code: 500, request id: 166CE53CC6F77322, host id: ",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-5041h169y5edpn",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 781,
  "error": "InternalError: We encountered an internal error, please try again.: cause(malformed chunked encoding)\n\tstatus code: 500, request id: 166CE53CC6F77322, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go PUT expected to success but got InternalError: We encountered an internal error, please try again.: cause(malformed chunked encoding)\n\tstatus code: 500, request id: 166CE53CC6F77322, host id: ",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 4 seconds
{
  "name": "aws-sdk-ruby",
  "function": "putObject(bucket_name, file)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-37a12be692cd",
    "file": "/mint/data/datafile-1-kB"
  },
  "duration": 2444.57,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 12 seconds
{
  "name": "awscli",
  "duration": 8261,
  "function": "aws --endpoint-url http://minio-c2.minio.io:30131 s3api put-object --body /mint/data/datafile-1-kB --bucket awscli-mint-test-bucket-26527 --key datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (InternalError) when calling the PutObject operation (reached max retries: 4): We encountered an internal error, please try again.: cause(malformed chunked encoding)"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 6 seconds
{
  "name": "mc",
  "duration": "4748",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-6109 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-11546/mc-test-object-14584  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-11546/mc-test-object-14584`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Object storage backend is unreachable\nTotal: 0 B, Transferred: 3.44 MiB, Speed: 770.65 KiB/s"
}
(3/15) Running minio-dotnet tests ... FAILED in 14 seconds
{
  "name": "minio-dotnet : RemoveObject_Test1",
  "function": "Task RemoveObjectAsync(string bucketName, string objectName, CancellationToken cancellationToken = default(CancellationToken))",
  "args": {
    "bucketName": "miniodotnet244uz1idgkznhaa",
    "objectName": "]}bcb#a$@d"
  },
  "duration": 145,
  "status": "FAIL",
  "alert": "",
  "message": "MinIO API responded with message=We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "error": "Minio.Exceptions.UnexpectedMinioException: MinIO API responded with message=We encountered an internal error, please try again.: cause(malformed chunked encoding)\n   at Minio.MinioClient.ParseErrorFromContent(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 538\n   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 431\n   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 605\n   at Minio.MinioClient.ExecuteTaskCoreAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 405\n   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, Dictionary`2 sseHeaders, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 532\n   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, ServerSideEncryption sse, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 301\n   at Minio.Functional.Tests.FunctionalTest.RemoveObject_Test1(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2013"
}
(3/15) Running minio-go tests ... FAILED in 5 seconds
{
  "args": {
    "bucketName": "minio-go-test-9rdbhmsr6kuhfsp5",
    "objectName": "2dlb6rxhvrxb0hf3mup6tr2cwik2b3",
    "region": "eu-west-1"
  },
  "duration": 4579,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "MakeBucket(bucketName, region)",
  "message": "PutObject failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket]",
  "duration": 80,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InternalError, message = We encountered an internal error, please try again.: cause(malformed chunked encoding), bucketName = minio-java-test-15jhhqk, objectName = minio-java-test-9voh4l, resource = /minio-java-test-15jhhqk/minio-java-test-9voh4l, requestId = 166CE54733BCE9C0, hostId = ddeb0b25-79fc-4fb8-9e0a-af404435c86a)\nrequest={method=PUT, url=http://minio-c2.minio.io:30131/minio-java-test-15jhhqk/minio-java-test-9voh4l, headers=Content-Type: application/octet-stream\nHost: minio-c2.minio.io:30131\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: yB5yjZ1ML2NvBn+JzBSGLA==\nx-amz-content-sha256: d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f90da3a666eec13ab35\nx-amz-date: 20210316T181420Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210316/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Length: 426\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166CE54733BCE9C0\nX-Xss-Protection: 1; mode=block\nDate: Tue, 16 Mar 2021 18:14:20 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.createObjects(FunctionalTest.java:1344), FunctionalTest.testListObjects(FunctionalTest.java:1395), FunctionalTest.listObjects(FunctionalTest.java:1428), FunctionalTest.runObjectTests(FunctionalTest.java:3747), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(3/15) Running minio-js tests ... FAILED in 1 seconds
{
  "name": "minio-js",
  "function": "putObject(bucketName, objectName, stream)",
  "args": "bucketName:minio-js-test-e48894b9-8b84-45b0-9876-d2953bc497ff, objectName:path/to/datafile-100-kB.buffer, stream:100Kib",
  "duration": 148,
  "status": "FAIL",
  "error": "S3Error: We encountered an internal error, please try again.: cause(malformed chunked encoding) at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(3/15) Running minio-py tests ... FAILED in 10 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-b69bb1df-1682-425d-b258-fab125ba3cc8",
    "object_name": "3ccaec13-aedd-481e-853a-891dcb208a39-f",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "HTTPConnectionPool(host='minio-c2.minio.io', port=30131): Max retries exceeded with url: /minio-py-test-b69bb1df-1682-425d-b258-fab125ba3cc8/3ccaec13-aedd-481e-853a-891dcb208a39-f (Caused by ResponseError('too many 500 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 334, in test_fput_object_small_file\n    _test_fput_object(bucket_name, object_name, _TEST_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 308, in _test_fput_object\n    _CLIENT.fput_object(bucket_name, object_name, filename, sse=sse)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1010, in fput_object\n    tags=tags, retention=retention, legal_hold=legal_hold,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 278, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 573, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-c2.minio.io', port=30131): Max retries exceeded with url: /minio-py-test-b69bb1df-1682-425d-b258-fab125ba3cc8/3ccaec13-aedd-481e-853a-891dcb208a39-f (Caused by ResponseError('too many 500 error responses',))\n",
  "duration": 9415
}
(3/15) Running s3cmd tests ... FAILED in 50 seconds
{
  "name": "s3cmd",
  "duration": "49021",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "WARNING: Upload failed: /s3cmd-test-object-11614 (500 (InternalError): We encountered an internal error, please try again.: cause(malformed chunked encoding))\nWARNING: Waiting 3 sec...\nWARNING: Upload failed: /s3cmd-test-object-11614 (500 (InternalError): We encountered an internal error, please try again.: cause(malformed chunked encoding))\nWARNING: Waiting 6 sec...\nWARNING: Upload failed: /s3cmd-test-object-11614 (500 (InternalError): We encountered an internal error, please try again.: cause(malformed chunked encoding))\nWARNING: Waiting 9 sec...\nWARNING: Upload failed: /s3cmd-test-object-11614 (500 (InternalError): We encountered an internal error, please try again.: cause(malformed chunked encoding))\nWARNING: Waiting 12 sec...\nWARNING: Upload failed: /s3cmd-test-object-11614 (500 (InternalError): We encountered an internal error, please try again.: cause(malformed chunked encoding))\nWARNING: Waiting 15 sec...\nERROR: Upload of '/mint/data/datafile-1-MB' failed too many times (Last reason: Upload failed for: /s3cmd-test-object-11614)"
}
(3/15) Running s3select tests ... FAILED in 13 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-a3fb6c50-c736-4691-846c-dfa6d3dc27c2"
  },
  "duration": 13234,
  "message": "Test test_1 unexpectedly failed with: HTTPConnectionPool(host='minio-c2.minio.io', port=30131): Max retries exceeded with url: /s3select-test-a3fb6c50-c736-4691-846c-dfa6d3dc27c2/4aa92a89-111a-4db3-9002-34e054e69db3 (Caused by ResponseError('too many 500 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 38, in test_sql_api\n    io.BytesIO(input_data), len(input_data))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 278, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 573, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-c2.minio.io', port=30131): Max retries exceeded with url: /s3select-test-a3fb6c50-c736-4691-846c-dfa6d3dc27c2/4aa92a89-111a-4db3-9002-34e054e69db3 (Caused by ResponseError('too many 500 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 118, in test_csv_input_custom_quote_char\n    data, sql_opts, expected_output)\n  File \"/mint/run/core/s3select/csv.py\", line 48, in test_sql_api\n    'Test {} unexpectedly failed with: {}'.format(test_name, select_err))\nValueError: Test test_1 unexpectedly failed with: HTTPConnectionPool(host='minio-c2.minio.io', port=30131): Max retries exceeded with url: /s3select-test-a3fb6c50-c736-4691-846c-dfa6d3dc27c2/4aa92a89-111a-4db3-9002-34e054e69db3 (Caused by ResponseError('too many 500 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running security tests ... done in 0 seconds

Executed 3 out of 15 tests successfully.
```
<a name="a1b37cd3-9e60-4d14-afa8-a9dfb89a83ba"></a>
### 11801-394cc2c/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:31723
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c26bd26e04c0:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 15 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 7 seconds
{
  "name": "mc",
  "duration": "5910",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-19674 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-17392/mc-test-object-18745  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-17392/mc-test-object-18745`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. We encountered an internal error, please try again.: cause(malformed chunked encoding)\nTotal: 0 B, Transferred: 9.06 MiB, Speed: 1.58 MiB/s"
}
(7/15) Running minio-dotnet tests ... done in 41 seconds
(8/15) Running minio-go tests ... FAILED in 23 seconds
{
  "args": {
    "bucketName": "minio-go-test-ufn4balwaspgxkvp",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 3082,
  "error": "We encountered an internal error, please try again.: cause(malformed chunked encoding)",
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "PutObjectStreaming failed",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 1 minutes and 19 seconds
(9/15) Running minio-js tests ... done in 47 seconds
(10/15) Running minio-py tests ... done in 3 minutes and 21 seconds
(11/15) Running s3cmd tests ... done in 18 seconds
(12/15) Running s3select tests ... done in 7 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/streaming-signature-v4.go
+++ b/cmd/streaming-signature-v4.go
@@ -166,72 +166,23 @@ func newSignV4ChunkedReader(req *http.Request) (io.ReadCloser, APIErrorCode) {
 		seedDate:          seedDate,
 		region:            region,
 		chunkSHA256Writer: sha256.New(),
-		state:             readChunkHeader,
+		buffer:            make([]byte, 64*1024),
 	}, ErrNone
 }
 
 // Represents the overall state that is required for decoding a
 // AWS Signature V4 chunked reader.
 type s3ChunkedReader struct {
-	reader            *bufio.Reader
-	cred              auth.Credentials
-	seedSignature     string
-	seedDate          time.Time
-	region            string
-	state             chunkState
-	lastChunk         bool
-	chunkSignature    string
+	reader        *bufio.Reader
+	cred          auth.Credentials
+	seedSignature string
+	seedDate      time.Time
+	region        string
+
 	chunkSHA256Writer hash.Hash // Calculates sha256 of chunk data.
-	n                 uint64    // Unread bytes in chunk
+	buffer            []byte
+	offset            int
 	err               error
-}
-
-// Read chunk reads the chunk token signature portion.
-func (cr *s3ChunkedReader) readS3ChunkHeader() {
-	// Read the first chunk line until CRLF.
-	var hexChunkSize, hexChunkSignature []byte
-	hexChunkSize, hexChunkSignature, cr.err = readChunkLine(cr.reader)
-	if cr.err != nil {
-		return
-	}
-	// <hex>;token=value - converts the hex into its uint64 form.
-	cr.n, cr.err = parseHexUint(hexChunkSize)
-	if cr.err != nil {
-		return
-	}
-	if cr.n == 0 {
-		cr.err = io.EOF
-	}
-	// Save the incoming chunk signature.
-	cr.chunkSignature = string(hexChunkSignature)
-}
-
-type chunkState int
-
-const (
-	readChunkHeader chunkState = iota
-	readChunkTrailer
-	readChunk
-	verifyChunk
-	eofChunk
-)
-
-func (cs chunkState) String() string {
-	stateString := ""
-	switch cs {
-	case readChunkHeader:
-		stateString = "readChunkHeader"
-	case readChunkTrailer:
-		stateString = "readChunkTrailer"
-	case readChunk:
-		stateString = "readChunk"
-	case verifyChunk:
-		stateString = "verifyChunk"
-	case eofChunk:
-		stateString = "eofChunk"
-
-	}
-	return stateString
 }
 
 func (cr *s3ChunkedReader) Close() (err error) {
@@ -241,83 +192,165 @@ func (cr *s3ChunkedReader) Close() (err error) {
 // Read - implements `io.Reader`, which transparently decodes
 // the incoming AWS Signature V4 streaming signature.
 func (cr *s3ChunkedReader) Read(buf []byte) (n int, err error) {
+	// First, if there is any unread data, copy it to the client
+	// provided buffer.
+	if cr.offset > 0 {
+		n = copy(buf, cr.buffer[cr.offset:])
+		if n == len(buf) {
+			cr.offset += n
+			return n, nil
+		}
+		cr.offset = 0
+		buf = buf[n:]
+	}
+
+	// Now, we read one chunk from the underlying reader.
+	// A chunk has the following format:
+	//   <chunk-size-as-hex> + ";chunk-signature=" + <signature-as-hex> + "\r\n" + <payload> + "\r\n"
+	//
+	// Frist, we read the chunk size but fail if it is larger
+	// than 1 MB. We must not accept arbitrary large chunks.
+	// One 1 MB is a reasonable max limit.
+	//
+	// Then we read the signature and payload data. We compute the SHA256 checksum
+	// of the payload and verify that it matches the expected signature value.
+	//
+	// The last chunk is *always* 0-sized. So, we must only return io.EOF if we have encountered
+	// a chunk with a chunk size = 0. However, this chunk still has a signature and we must
+	// verify it.
+	const MaxSize = 1 << 20 // 1 MB
+	var size int
 	for {
-		switch cr.state {
-		case readChunkHeader:
-			cr.readS3ChunkHeader()
-			// If we're at the end of a chunk.
-			if cr.n == 0 && cr.err == io.EOF {
-				cr.state = readChunkTrailer
-				cr.lastChunk = true
-				continue
-			}
-			if cr.err != nil {
-				return 0, cr.err
-			}
-			cr.state = readChunk
-		case readChunkTrailer:
-			cr.err = readCRLF(cr.reader)
-			if cr.err != nil {
-				return 0, errMalformedEncoding
-			}
-			cr.state = verifyChunk
-		case readChunk:
-			// There is no more space left in the request buffer.
-			if len(buf) == 0 {
-				return n, nil
-			}
-			rbuf := buf
-			// The request buffer is larger than the current chunk size.
-			// Read only the current chunk from the underlying reader.
-			if uint64(len(rbuf)) > cr.n {
-				rbuf = rbuf[:cr.n]
-			}
-			var n0 int
-			n0, cr.err = cr.reader.Read(rbuf)
-			if cr.err != nil {
-				// We have lesser than chunk size advertised in chunkHeader, this is 'unexpected'.
-				if cr.err == io.EOF {
-					cr.err = io.ErrUnexpectedEOF
-				}
-				return 0, cr.err
-			}
+		b, err := cr.reader.ReadByte()
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+		if err != nil {
+			cr.err = err
+			return n, cr.err
+		}
+		if b == ';' { // separating character
+			break
+		}
 
-			// Calculate sha256.
-			cr.chunkSHA256Writer.Write(rbuf[:n0])
-			// Update the bytes read into request buffer so far.
-			n += n0
-			buf = buf[n0:]
-			// Update bytes to be read of the current chunk before verifying chunk's signature.
-			cr.n -= uint64(n0)
-
-			// If we're at the end of a chunk.
-			if cr.n == 0 {
-				cr.state = readChunkTrailer
-				continue
-			}
-		case verifyChunk:
-			// Calculate the hashed chunk.
-			hashedChunk := hex.EncodeToString(cr.chunkSHA256Writer.Sum(nil))
-			// Calculate the chunk signature.
-			newSignature := getChunkSignature(cr.cred, cr.seedSignature, cr.region, cr.seedDate, hashedChunk)
-			if !compareSignatureV4(cr.chunkSignature, newSignature) {
-				// Chunk signature doesn't match we return signature does not match.
-				cr.err = errSignatureMismatch
-				return 0, cr.err
-			}
-			// Newly calculated signature becomes the seed for the next chunk
-			// this follows the chaining.
-			cr.seedSignature = newSignature
-			cr.chunkSHA256Writer.Reset()
-			if cr.lastChunk {
-				cr.state = eofChunk
-			} else {
-				cr.state = readChunkHeader
-			}
-		case eofChunk:
-			return n, io.EOF
+		// Manually deserialize the size since AWS specified
+		// the chunk size to be of variable width. In particular,
+		// a size of 16 is encoded as `10` while a size of 64 KB
+		// is `10000`.
+		switch {
+		case b >= '0' && b <= '9':
+			size = size<<4 | int(b-'0')
+		case b >= 'a' && b <= 'f':
+			size = size<<4 | int(b-('a'-10))
+		case b >= 'A' && b <= 'F':
+			size = size<<4 | int(b-('A'-10))
+		default:
+			cr.err = errMalformedEncoding
+			return n, cr.err
+		}
+		if size > MaxSize {
+			cr.err = errMalformedEncoding
+			return n, cr.err
 		}
 	}
+
+	// Now, we read the signature of the following payload and expect:
+	//   chunk-signature=" + <signature-as-hex> + "\r\n"
+	//
+	// The signature is 64 bytes long (hex-encoded SHA256 hash) and
+	// starts with a 16 byte header: len("chunk-signature=") + 64 == 80.
+	var signature [80]byte
+	_, err = io.ReadFull(cr.reader, signature[:])
+	if err == io.EOF {
+		err = io.ErrUnexpectedEOF
+	}
+	if err != nil {
+		cr.err = err
+		return n, cr.err
+	}
+	if !bytes.HasPrefix(signature[:], []byte("chunk-signature=")) {
+		cr.err = errMalformedEncoding
+		return n, cr.err
+	}
+	b, err := cr.reader.ReadByte()
+	if err == io.EOF {
+		err = io.ErrUnexpectedEOF
+	}
+	if err != nil {
+		cr.err = err
+		return n, cr.err
+	}
+	if b != '\r' {
+		cr.err = errMalformedEncoding
+		return n, cr.err
+	}
+	b, err = cr.reader.ReadByte()
+	if err == io.EOF {
+		err = io.ErrUnexpectedEOF
+	}
+	if err != nil {
+		cr.err = err
+		return n, cr.err
+	}
+	if b != '\n' {
+		cr.err = errMalformedEncoding
+		return n, cr.err
+	}
+
+	if cap(cr.buffer) < size {
+		cr.buffer = make([]byte, size)
+	} else {
+		cr.buffer = cr.buffer[:size]
+	}
+
+	// Now, we read the payload and compute its SHA-256 hash.
+	_, err = io.ReadFull(cr.reader, cr.buffer)
+	if err == io.EOF && size != 0 {
+		err = io.ErrUnexpectedEOF
+	}
+	if err != nil && err != io.EOF {
+		cr.err = err
+		return n, cr.err
+	}
+	b, err = cr.reader.ReadByte()
+	if b != '\r' {
+		cr.err = errMalformedEncoding
+		return n, cr.err
+	}
+	b, err = cr.reader.ReadByte()
+	if err == io.EOF {
+		err = io.ErrUnexpectedEOF
+	}
+	if err != nil {
+		cr.err = err
+		return n, cr.err
+	}
+	if b != '\n' {
+		cr.err = errMalformedEncoding
+		return n, cr.err
+	}
+
+	// Once we have read the entire chunk successfully, we verify
+	// that the received signature matches our computed signature.
+	cr.chunkSHA256Writer.Write(cr.buffer)
+	newSignature := getChunkSignature(cr.cred, cr.seedSignature, cr.region, cr.seedDate, hex.EncodeToString(cr.chunkSHA256Writer.Sum(nil)))
+	if !compareSignatureV4(string(signature[16:]), newSignature) {
+		cr.err = errSignatureMismatch
+		return n, cr.err
+	}
+	cr.seedSignature = newSignature
+	cr.chunkSHA256Writer.Reset()
+
+	// If the chunk size is zero we return io.EOF. As specified by AWS,
+	// only the last chunk is zero-sized.
+	if size == 0 {
+		cr.err = io.EOF
+		return n, cr.err
+	}
+
+	cr.offset = copy(buf, cr.buffer)
+	n += cr.offset
+	return n, err
 }
 
 // readCRLF - check if reader only has '\r\n' CRLF character.

